### PR TITLE
Adding support for unaligned load and store for WebAssembly

### PR DESCRIPTION
**Pull request by arunetm**
Adding new opcodes for WASM load/store allowing unaligned access.
Adding Interpreter and JIT support for new operands.
Adding tests.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/1494)

### CODE REVIEW

**Comment by arunetm** [comment]
@Cellule @MikeHolman review please.

**Comment by Cellule** [comment]
Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[lib/Runtime/ByteCode/AsmJsByteCodeDumper.cpp, line 554 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3JQwVkNPJ2e7OZzq8:-KQ3JQwVkNPJ2e7OZzq9:bsqotc7) ([raw file](https://github.com/microsoft/chakracore/blob/6aa127ef7fac5f364917cb129d11cfb8172a1dce/lib/Runtime/ByteCode/AsmJsByteCodeDumper.cpp#L554)):*
> ```cpp
>         case OpCodeAsmJs::LdArr:
>             Output::Print(_u("%c%d = %s[I%d]"), valueTag, data->Value, heapTag, data->SlotIndex); break;
>         case OpCodeAsmJs::LdArrWasm:
> ```

I don't think it has much value to have a different output since we know we're in a wasm function anyway.

---

*[test/wasm/address.wast, line 1 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3KaqafmVZg9iEZP3N:-KQ3KaqafmVZg9iEZP3O:b20vtb2) ([raw file](https://github.com/microsoft/chakracore/blob/6aa127ef7fac5f364917cb129d11cfb8172a1dce/test/wasm/address.wast#L1)):*
> ```
> ;;-------------------------------------------------------------------------------------------------------
> ```

Don't we already have that exact same test in WasmSpec ?

---

*[test/wasm/f32address.wast, line 1 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3L1jtxgl7LZXnXCRs:-KQ3L1jtxgl7LZXnXCRt:b4q6wvo) ([raw file](https://github.com/microsoft/chakracore/blob/6aa127ef7fac5f364917cb129d11cfb8172a1dce/test/wasm/f32address.wast#L1)):*
> ```
> ;;-------------------------------------------------------------------------------------------------------
> ```

Don't we already have those tests in WasmSpec ?

---

*[test/WasmSpec/baselines/float_memory.baseline, line 7 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3LA36i_OpeZ20BGpn:-KQ3LA36i_OpeZ20BGpo:b-8bexlg) ([raw file](https://github.com/microsoft/chakracore/blob/e3dd7a4b8a777dc6b09482203e5c1f8da98d7393/test/WasmSpec/baselines/float_memory.baseline#L7)):*
> ```
> Unexpected Error while compiling float_memory.1.wasm
> Error: Compiling wasm failed: I64 support NYI
> float_memory.wast:95: $assert_return_0 failed.
> ```

Cheers !

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1494)*

File: test/WasmSpec/testsuite/float_memory.wast [hunk 1]
```diff
@@ -92,7 +92,7 @@
 )
 
 ;;(assert_return (invoke "i32.load") (i32.const 0x7fa00000))
-(assert_return (invoke "f32.load") (f32.const nan:0x200000))
+;;(assert_return (invoke "f32.load") (f32.const nan:0x200000))
```
Comment (arunetm) [inline test/WasmSpec/testsuite/float_memory.wast]: @MikeHolman, This test validates the load of a wasm canonical NaN. The test fails the assertion in native execution as we do not store the NaN in WASM canonical form even though the argument is canonical. (but chakra perserves the bits during interpretation and passes the test). 
Enforcing NaN canonicalization in WASM solves this issue. I have a prototype fix for this, but the ideal solution is to implement the NaN canonicalization for all float operations as recommended by the wasm design. 

Do you have any thoughts on proceeding with enforcing WASM NaN canonicalization for the f32 operations at this point?

Disabling this test for the time being.

Comment (MikeHolman) [inline test/WasmSpec/testsuite/float_memory.wast]: i'm not worried about exactly correct nan canonicalization right now. I think it's fine to disable for the time being.

**Comment by arunetm** [comment]
*[test/wasm/f32address.wast, line 1 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3L1jtxgl7LZXnXCRs:-KQIzc6dRSggfRs-liGC:bamy7ap) ([raw file](https://github.com/microsoft/chakracore/blob/6aa127ef7fac5f364917cb129d11cfb8172a1dce/test/wasm/f32address.wast#L1)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary>
> Don't we already have those tests in WasmSpec ?
</details>
This test is different. WasmSpec tests validates nan canonicalization for these float tests. We dont canonicalize nan's yet. 
This test doesnt use canonical nan's but validates the load/store behavior for float32

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1494)*

**Comment by arunetm** [comment]
*[lib/Runtime/ByteCode/AsmJsByteCodeDumper.cpp, line 554 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3JQwVkNPJ2e7OZzq8:-KQJ-C45nf_MPKVIQV6W:bqitkd3) ([raw file](https://github.com/microsoft/chakracore/blob/6aa127ef7fac5f364917cb129d11cfb8172a1dce/lib/Runtime/ByteCode/AsmJsByteCodeDumper.cpp#L554)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary>
> I don't think it has much value to have a different output since we know we're in a wasm function anyway.
</details>
Yes I agree. This is not critical to distinguish the wasm opcodes in the dumper.  Will remove this.
The only benefit i can think of is making it explicit to the developer that we are in a wasm module.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1494)*

**Comment by arunetm** [comment]
*[test/wasm/address.wast, line 1 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3KaqafmVZg9iEZP3N:-KQJ-iewApLHE9NrqNFS:bkkfmaz) ([raw file](https://github.com/microsoft/chakracore/blob/6aa127ef7fac5f364917cb129d11cfb8172a1dce/test/wasm/address.wast#L1)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary>
> Don't we already have that exact same test in WasmSpec ?
</details>
yes we do, I will remove this test and enable it through the WasmSpec tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1494)*

**Comment by arunetm** [comment]
*[test/WasmSpec/testsuite/float_memory.wast, line 95 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/1494#-KQ3JOBBRCd7TGWlj8yV-r1-95:-KQJ01TaEUZEQg8UzVC8:bau6h4i) ([raw file](https://github.com/microsoft/chakracore/blob/6aa127ef7fac5f364917cb129d11cfb8172a1dce/test/WasmSpec/testsuite/float_memory.wast#L95)):*
<details><summary><i>Previously, MikeHolman (Michael Holman) wrote…</i></summary>
> i'm not worried about exactly correct nan canonicalization right now. I think it's fine to disable for the time being.
</details>
Thanks. Disabling the test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1494)*

**Comment by arunetm** [comment]
@MikeHolman can you please take a look?

**Comment by MikeHolman** [comment]
:shipit:

**Comment by arunetm** [comment]
Thanks, Squashed into a single commit and re-signed the changes to prepare for merge.

**Pull request merged**